src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+                                           MacroAssembler *_masm) {
+
+  assert_different_registers(eq_mask, haystack, needleLen, rTmp, hsLen, r10);
+  assert(r10 == nMinusK, "Bad register");
+
+  const XMMRegister byte_0 = XMM_BYTE_0;
+  const XMMRegister byte_1 = XMM_BYTE_1;
+  const XMMRegister byte_2 = XMM_BYTE_2;
+  const XMMRegister byte_k = XMM_BYTE_K;
+
+  int scale = isU ? 2 : 1;
+  std::function<void(XMMRegister dst, XMMRegister src, Address adr, int vector_len)> vpcmpeq;
+
+  if (isU) {
+    vpcmpeq = [_masm](XMMRegister dst, XMMRegister src, Address adr, int vector_len) {
+      __ vpcmpeqw(dst, src, adr, vector_len);
+    };
+  } else {
+    vpcmpeq = [_masm](XMMRegister dst, XMMRegister src, Address adr, int vector_len) {
+      __ vpcmpeqb(dst, src, adr, vector_len);
+    };
+  }
+
+#undef cmp_0
+#define cmp_0 XMM_TMP3
+#undef cmp_k
+#define cmp_k XMM_TMP4
+#undef lastCompare
+#define lastCompare rTmp
+#undef lastMask
+
+  int sizeIncr = isU ? 2 : 1;
+  bool needToSaveRCX = false;
+
+  needToSaveRCX = (haystack == rcx) || (hsLen == rcx) || (eq_mask == rcx) || (rTmp == rcx) || (needleLen == rcx)
+                  || (nMinusK == rcx);
+
+  Label L_OKtoCompareFull, L_done, L_specialCase_gt2, L_specialCase_gt3;
+
+  assert(!sizeKnown || (sizeKnown && ((size > 0) && (size <= NUMBER_OF_CASES))), "Incorrect size given");
+
+  Address kThByte
+      = sizeKnown ? Address(haystack, size - sizeIncr) : Address(haystack, needleLen, Address::times_1, -(sizeIncr));
+  size = sizeKnown ? size : NUMBER_OF_CASES + 1;
+
+  // Compare first byte of needle to haystack
+     vpcmpeq(cmp_0, byte_0, Address(haystack, 0), Assembler::AVX_256bit);
+
+  if (doEarlyBailout) {
+    __ vpmovmskb(eq_mask, cmp_0, Assembler::AVX_256bit);
+    __ testl(eq_mask, eq_mask);
+    __ je(noMatch);
+  }
+
+  if (size == (isU ? 2 : 1)) {
+    __ vpmovmskb(eq_mask, cmp_0, Assembler::AVX_256bit);
+  } else {
+    __ cmpq(nMinusK, 32);
+    __ jae(L_OKtoCompareFull);
+    __ vpmovmskb(eq_mask, cmp_0, Assembler::AVX_256bit);
+
+    // If n-k less than 32, comparing the last byte of the needle will result
+    // in reading past the end of the haystack.  Account for this here.
+    __ leaq(lastCompare, Address(haystack, hsLen, Address::times_1, -32));
+#undef cmp_0
+#undef saveRCX
+#define saveRCX XMM_TMP3
+    if (needToSaveRCX) {
+      __ movdq(saveRCX, rcx);
+    }
+#undef shiftVal
+#define shiftVal rcx
+    __ movq(shiftVal, isU ? 30 : 31);
+    __ subq(shiftVal, nMinusK);
+
+       vpcmpeq(cmp_k, byte_k, Address(lastCompare, 0), Assembler::AVX_256bit);
+
+#undef lastCompare
+#define lastMask rTmp
+    __ vpmovmskb(lastMask, cmp_k, Assembler::AVX_256bit);
+    __ shrq(lastMask);
+#undef shiftVal
+    __ andq(eq_mask, lastMask);
+    if (needToSaveRCX) {
+      __ movdq(rcx, saveRCX);
+    }
+#undef saveRCX
+
+    if (bytesToCompare > 2) {
+      if (size > (isU ? 4 : 2)) {
+        if (doEarlyBailout) {
+          __ testl(eq_mask, eq_mask);
```

Comment:
The masks are 32 bit as we are comparing max 32 byes (256 bits) at a time. So we could consistently do either andl, testl, shrl or andq, testq, shrq.